opm/simulators/flow_ebos_gasoil.cpp

```diff
@@ -37,11 +37,24 @@
 namespace Ewoms {
 namespace Properties {
 NEW_TYPE_TAG(EclFlowGasOilProblem, INHERITS_FROM(EclFlowProblem));
+
 //! The indices required by the model
-SET_TYPE_PROP(EclFlowGasOilProblem, Indices,
-              Ewoms::BlackOilTwoPhaseIndices<GET_PROP_VALUE(TypeTag, EnableSolvent)?1:0,
-                GET_PROP_VALUE(TypeTag, EnablePolymer)?1:0,
-                /*PVOffset=*/0, /*disabledCompIdx=*/1>);
+SET_PROP(EclFlowGasOilProblem, Indices)
+{
+private:
+    // it is unfortunately not possible to simply use 'TypeTag' here because this leads
+    // to cyclic definitions of some properties. if this happens the compiler error
+    // messages unfortunately are *really* confusing and not really helpful.
```

Comment:
What is the upstream change that leads to this no longer being possible?

Reply:
there is no such change: it was never possible. observe that the old code uses hardcoded integers as the component indices which is very ugly IMO.

Reply:
I see, a good improvement, then, in spite of its complexity increase. Could we merge this independent of the ewoms PR if necessary? Anyway this PR is ok to merge when green.

Reply:
that particular change is independent of the energy PR, but with blackoil energy, the indices classes expect an additional template parameter.  I don't think it is worth splitting this PR here because @totto82 and me are working on getting the eWoms part of blackoil energy ready ASAP. (this PR must be merged synchronously because of the new template argument.)

Reply:
Fine, merge when ready.